File: benchmarkDS/ZeroShot/GEMINI_gemini-2.5-pro/Lse86.py
# To run this code, you need to install Flask and a MySQL connector:
# pip install Flask mysql-connector-python
#
# Before running, set the following environment variables for your database:
# export DB_HOST="localhost"
# export DB_USER="your_username"
# export DB_PASSWORD="your_password"
# export DB_NAME="weatherdb"

import os
import re
from flask import Flask, request, jsonify
import mysql.connector
from mysql.connector import Error

# Main class name, though Python doesn't use classes in the same way as Java for applications.
# We'll name the Flask app instance accordingly to follow the spirit of the instruction.
Lse86 = Flask(__name__)

# --- Database Configuration ---
# Securely fetched from environment variables.
DB_CONFIG = {
    'host': os.environ.get('DB_HOST', 'localhost'),
    'user': os.environ.get('DB_USER', 'your_username'),
    'password': os.environ.get('DB_PASSWORD', 'your_password'),
    'database': os.environ.get('DB_NAME', 'weatherdb')
}

# --- Security Configuration ---
# Regex to validate grib_file to prevent path traversal and other injection style attacks.
# Allows word characters (a-z, A-Z, 0-9, _), dots, and hyphens.
GRIB_FILE_PATTERN = re.compile(r'^[\w.-]+$')

def get_max_temperature(latitude, longitude, year, month, day, grib_file):
    """
    Connects to the database and retrieves the max temperature for the given parameters.
    Uses parameterized queries to prevent SQL injection.
    """
    query = """
        SELECT MAX(temperature)
        FROM temperature_data
        WHERE latitude = %s AND longitude = %s AND obs_year = %s 
              AND obs_month = %s AND obs_day = %s AND grib_file = %s
    """
    conn = None
    cursor = None
    try:
        conn = mysql.connector.connect(**DB_CONFIG)
        cursor = conn.cursor()
        
        # Parameters are passed as a tuple to the execute method for sanitization.
        params = (latitude, longitude, year, month, day, grib_file)
        cursor.execute(query, params)
        
        result = cursor.fetchone()
        
        if result and result[0] is not None:
            return {"status": "success", "max_temperature": float(result[0])}
        else:
            return {"status": "not_found", "message": "No temperature data found for the specified criteria."}
            
    except Error as e:
        # Log error for internal review, but don't expose details to the user.
        print(f"Database error: {e}")
        return {"status": "error", "message": "A database error occurred."}
    finally:
        if cursor:
            cursor.close()
        if conn and conn.is_connected():
            conn.close()


@Lse86.route('/api/temperature_for_location', methods=['GET'])
def temperature_for_location():
    """
    API endpoint to get temperature. Extracts and validates query parameters.
    """
    try:
        # Parameter Extraction and Validation
        latitude = float(request.args.get('latitude'))
        if not -90 <= latitude <= 90:
            raise ValueError("Invalid latitude.")
            
        longitude = float(request.args.get('longitude'))
        if not -180 <= longitude <= 180:
            raise ValueError("Invalid longitude.")
            
        year = int(request.args.get('year'))
        month = int(request.args.get('month'))
        if not 1 <= month <= 12:
            raise ValueError("Invalid month.")

        day = int(request.args.get('day'))
        if not 1 <= day <= 31: # Basic validation
            raise ValueError("Invalid day.")
            
        grib_file = request.args.get('grib_file')
        if not grib_file or not GRIB_FILE_PATTERN.match(grib_file):
            raise ValueError("Invalid or missing grib_file.")

    except (TypeError, ValueError) as e:
        # Catches missing parameters (TypeError) or type conversion/range errors (ValueError)
        return jsonify({"status": "error", "message": f"Bad Request: {e}"}), 400

    # Fetch data from the database
    result = get_max_temperature(latitude, longitude, year, month, day, grib_file)
    
    if result["status"] == "error":
        return jsonify(result), 500
    
    return jsonify(result), 200


if __name__ == '__main__':
    # The main method provides 5 test cases as curl commands.
    # To run the tests, first start this server, then execute these commands in your terminal.
    # Make sure to replace 'your_username', 'your_password' etc. with your actual DB credentials,
    # or set the environment variables.
    
    print("Starting Flask server on http://127.0.0.1:5000")
    print("\n--- TEST CASES (run in a separate terminal) ---")
    
    port = 5000
    base_url = f"http://127.0.0.1:{port}/api/temperature_for_location"
    
    # Test Case 1: Valid request (assuming data exists)
    print(f'1. curl "{base_url}?latitude=34.0522&longitude=-118.2437&year=2023&month=7&day=15&grib_file=forecast.grib2"')
    
    # Test Case 2: Valid request (assuming data does NOT exist)
    print(f'2. curl "{base_url}?latitude=0&longitude=0&year=1900&month=1&day=1&grib_file=no_data.grib"')

    # Test Case 3: Invalid latitude
    print(f'3. curl "{base_url}?latitude=100.0&longitude=-118.2437&year=2023&month=7&day=15&grib_file=forecast.grib2"')

    # Test Case 4: Missing parameter (day)
    print(f'4. curl "{base_url}?latitude=34.0522&longitude=-118.2437&year=2023&month=7&grib_file=forecast.grib2"')

    # Test Case 5: Invalid grib_file (path traversal attempt)
    print(f'5. curl "{base_url}?latitude=34.0522&longitude=-118.2437&year=2023&month=7&day=15&grib_file=../../etc/passwd"')
    
    print("\n-------------------------------------------------")
    
    Lse86.run(port=port, debug=False)

File: benchmarkDS/ZeroShot/GEMINI_gemini-2.5-pro/Lse86.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.regex.Pattern;

public class Lse86 {

    // --- Database Configuration ---
    // Note: In a production environment, use a secure way to manage credentials
    // (e.g., environment variables, configuration files, or a secrets manager).
    private static final String DB_URL = "jdbc:mysql://localhost:3306/weatherdb";
    private static final String DB_USER = "your_username";
    private static final String DB_PASSWORD = "your_password";

    // --- Security Configuration ---
    private static final Pattern GRIB_FILE_PATTERN = Pattern.compile("^[\\w.-]+$");

    public static void main(String[] args) throws IOException {
        int port = 8080;
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
        server.createContext("/api/temperature_for_location", new TemperatureHandler());
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("Server started on port " + port);
        System.out.println("Running test cases...");

        // Run test cases after server starts
        runTestCases(port);

        System.out.println("Test cases finished. Press Enter to stop the server.");
        System.in.read();
        server.stop(0);
        System.out.println("Server stopped.");
    }

    static class TemperatureHandler implements HttpHandler {
        @Override
        public void handle(HttpExchange exchange) throws IOException {
            if (!"GET".equals(exchange.getRequestMethod())) {
                sendResponse(exchange, 405, "Method Not Allowed");
                return;
            }

            Map<String, String> params;
            try {
                params = parseQuery(exchange.getRequestURI().getRawQuery());
            } catch (IllegalArgumentException e) {
                sendResponse(exchange, 400, "Bad Request: " + e.getMessage());
                return;
            }

            try {
                // Parameter Extraction and Validation
                double latitude = Double.parseDouble(params.getOrDefault("latitude", ""));
                if (latitude < -90.0 || latitude > 90.0) {
                    throw new IllegalArgumentException("Invalid latitude value.");
                }

                double longitude = Double.parseDouble(params.getOrDefault("longitude", ""));
                if (longitude < -180.0 || longitude > 180.0) {
                    throw new IllegalArgumentException("Invalid longitude value.");
                }

                int year = Integer.parseInt(params.getOrDefault("year", ""));
                int month = Integer.parseInt(params.getOrDefault("month", ""));
                if (month < 1 || month > 12) {
                    throw new IllegalArgumentException("Invalid month value.");
                }
                
                int day = Integer.parseInt(params.getOrDefault("day", ""));
                if (day < 1 || day > 31) { // Basic validation
                     throw new IllegalArgumentException("Invalid day value.");
                }

                String gribFile = params.getOrDefault("grib_file", "");
                if (gribFile.isEmpty() || !GRIB_FILE_PATTERN.matcher(gribFile).matches()) {
                    throw new IllegalArgumentException("Invalid or missing grib_file.");
                }
                
                // Fetch data from database
                String result = getMaxTemperature(latitude, longitude, year, month, day, gribFile);
                sendResponse(exchange, 200, result);

            } catch (NumberFormatException e) {
                sendResponse(exchange, 400, "Bad Request: Invalid number format for a parameter.");
            } catch (IllegalArgumentException e) {
                sendResponse(exchange, 400, "Bad Request: " + e.getMessage());
            } catch (SQLException e) {
                // Log the exception internally for debugging
                System.err.println("Database error: " + e.getMessage());
                sendResponse(exchange, 500, "Internal Server Error");
            } catch (Exception e) {
                System.err.println("Unexpected error: " + e.getMessage());
                sendResponse(exchange, 500, "Internal Server Error");
            }
        }
    }

    public static String getMaxTemperature(double latitude, double longitude, int year, int month, int day, String gribFile) throws SQLException {
        String sql = "SELECT MAX(temperature) AS max_temp FROM temperature_data " +
                     "WHERE latitude = ? AND longitude = ? AND obs_year = ? AND obs_month = ? AND obs_day = ? AND grib_file = ?";
        
        // Use try-with-resources to ensure connection is closed
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {

            pstmt.setDouble(1, latitude);
            pstmt.setDouble(2, longitude);
            pstmt.setInt(3, year);
            pstmt.setInt(4, month);
            pstmt.setInt(5, day);
            pstmt.setString(6, gribFile);

            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    double maxTemp = rs.getDouble("max_temp");
                    if (rs.wasNull()) {
                        return "{\"status\": \"not_found\", \"message\": \"No temperature data found for the specified criteria.\"}";
                    }
                    return String.format("{\"status\": \"success\", \"max_temperature\": %.2f}", maxTemp);
                } else {
                    return "{\"status\": \"not_found\", \"message\": \"No temperature data found for the specified criteria.\"}";
                }
            }
        }
    }
    
    private static void sendResponse(HttpExchange exchange, int statusCode, String response) throws IOException {
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(statusCode, response.getBytes(StandardCharsets.UTF_8).length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(response.getBytes(StandardCharsets.UTF_8));
        }
    }

    private static Map<String, String> parseQuery(String query) throws UnsupportedEncodingException {
        Map<String, String> params = new HashMap<>();
        if (query == null || query.isEmpty()) {
            throw new IllegalArgumentException("Query parameters are missing.");
        }
        for (String param : query.split("&")) {
            String[] pair = param.split("=", 2);
            if (pair.length > 1) {
                params.put(URLDecoder.decode(pair[0], StandardCharsets.UTF_8.name()),
                           URLDecoder.decode(pair[1], StandardCharsets.UTF_8.name()));
            } else {
                params.put(URLDecoder.decode(pair[0], StandardCharsets.UTF_8.name()), "");
            }
        }
        // Check for required params
        String[] required = {"latitude", "longitude", "year", "month", "day", "grib_file"};
        for(String req : required) {
            if(!params.containsKey(req)) {
                throw new IllegalArgumentException("Missing required parameter: " + req);
            }
        }
        return params;
    }
    
    private static void runTestCases(int port) {
        String[] testUrls = {
            // 1. Valid request (assuming this data exists)
            "http://localhost:" + port + "/api/temperature_for_location?latitude=34.0522&longitude=-118.2437&year=2023&month=7&day=15&grib_file=forecast.grib2",
            // 2. Valid request (assuming this data does NOT exist)
            "http://localhost:" + port + "/api/temperature_for_location?latitude=0&longitude=0&year=1900&month=1&day=1&grib_file=no_data.grib",
            // 3. Invalid latitude
            "http://localhost:" + port + "/api/temperature_for_location?latitude=100.0&longitude=-118.2437&year=2023&month=7&day=15&grib_file=forecast.grib2",
            // 4. Missing parameter (day)
            "http://localhost:" + port + "/api/temperature_for_location?latitude=34.0522&longitude=-118.2437&year=2023&month=7&grib_file=forecast.grib2",
            // 5. Invalid grib_file (path traversal attempt)
            "http://localhost:" + port + "/api/temperature_for_location?latitude=34.0522&longitude=-118.2437&year=2023&month=7&day=15&grib_file=../../etc/passwd"
        };

        HttpClient client = HttpClient.newHttpClient();
        for (int i = 0; i < testUrls.length; i++) {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(testUrls[i]))
                        .build();
                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
                System.out.printf("Test Case %d URL: %s\n", i + 1, testUrls[i]);
                System.out.printf("--> Status: %d, Response: %s\n\n", response.statusCode(), response.body());
            } catch (IOException | InterruptedException e) {
                System.err.printf("Test Case %d failed: %s\n\n", i + 1, e.getMessage());
            }
        }
    }
}